### PR TITLE
Use ioctl RNDADDENTROPY for writing back random seed

### PR DESCRIPTION
**Pull request by bitstreamout**
as this also increases the entropy bit count.  Compare with linux/drivers/char/random.c

### CODE REVIEW

File: src/random-seed/random-seed.c [hunk 1]
```diff
@@ -66,11 +94,13 @@ int main(int argc, char *argv[]) {
         if (buf_size <= POOL_SIZE_MIN)
                 buf_size = POOL_SIZE_MIN;
 
-        buf = malloc(buf_size);
-        if (!buf) {
+        entropy = (struct rand_pool_info*) malloc(sizeof(struct rand_pool_info) + buf_size);
```
Comment (keszybz) [inline src/random-seed/random-seed.c]: This cast is not needed (malloc returns `void *`).

File: src/random-seed/random-seed.c [hunk 2]
```diff
@@ -66,11 +94,13 @@ int main(int argc, char *argv[]) {
...
+        if (!entropy) {
                 r = log_oom();
                 goto finish;
         }
+        entropy->buf_size = (typeof(entropy->buf_size)) buf_size;
```
Comment (keszybz) [inline src/random-seed/random-seed.c]: Please don't. entropy->buf_size is an int. Let's not try to hide this.
Removing the cast altogether doesn't result in a warning here, so it should just go away.

File: src/random-seed/random-seed.c [hunk 3]
```diff
@@ -113,7 +143,7 @@ int main(int argc, char *argv[]) {
                         }
                 }
 
-                k = loop_read(seed_fd, buf, buf_size, false);
+                k = loop_read(seed_fd, (void*)entropy, buf_size, false);
```
Comment (keszybz) [inline src/random-seed/random-seed.c]: Likewise, casting to `void*` is not useful.

File: src/random-seed/random-seed.c [hunk 4]
```diff
@@ -166,7 +196,7 @@ int main(int argc, char *argv[]) {
                         goto finish;
                 }
 
-                r = loop_write(seed_fd, buf, (size_t) k, false);
+                r = loop_write(seed_fd, (void*)entropy, (size_t) k, false);
```
Comment (keszybz) [inline src/random-seed/random-seed.c]: Ditto. You can remove `(size_t)` cast too.

File: src/random-seed/random-seed.c [hunk 5]
```diff
@@ -33,9 +35,35 @@
...
+        ssize_t len;
+        int ret;
+
+        ret = ioctl(fd, RNDGETENTCNT, &((*entropy)->entropy_count));
+        if (ret < 0)
+                return -errno;
+
+        len = loop_read(fd, (*entropy)->buf, (*entropy)->buf_size, false);
+        if (len < 0)
+                return -len;
```
Comment (keszybz) [inline src/random-seed/random-seed.c]: `return len;`

File: src/random-seed/random-seed.c [hunk 6]
```diff
@@ -33,9 +35,35 @@
 
 #define POOL_SIZE_MIN 512
 
+static ssize_t get_random(int fd, struct rand_pool_info **entropy) {
+        ssize_t len;
+        int ret;
```
Comment (keszybz) [inline src/random-seed/random-seed.c]: `int r` is customary in systemd (`ret` is used for output parameters).

**Comment by poettering** [comment]
Hmm, I am really not convinced adding this is a good idea, see the discussions on #4271. There's really no reason to believe that the random seed stored on disk actually should be adding entropy, as it is systematically incompatible with booting systems from identical images. Consider 100 systems booting up from one and the same image, which ships the same random-seed file (since it is the _same_ image after all), then after bootup there's a good chance that most of them came to the same pool state and think there was actually entropy in it, even though there really isn't as there a ton other systems with the very same pool state...

I don't see how this could be fixed. There's no way to figure out if you are booting from a freshly duplicated image (where you better shouldn't consider the seed to contain entropy) or from a system with a single linear line of clean boots (where it is OK to assume that the seed contains entropy).

I am thinking that this is not worth adding at all, since it creates a fake sense of entropy in a good part of systems (as pre-built images are kinda common these days – think containers –, and individually put together ones kinda old-school). What we maybe could offer is doing this optionally, via a kernel command line switch, but it really needs to be opt-in, not opt-out, and documentation must make clear it's incompatible with a pre-built image style setup.

**Comment by keszybz** [comment]
> Consider 100 systems booting up from one and the same image, which ships the same random-seed file

I believe that if you're making an image for duplication, things like random-seed, ssh keys, etc, should be nuked from the image.

**Comment by poettering** [comment]
> I believe that if you're making an image for duplication, things like random-seed, ssh keys, etc, should be nuked from the image.

Sure, but who actually does that? And if they do, who'd think about random-seed?

I mean, the difference between the random-seed on one hand and ssh keys or /etc/machine-id on the other in this regard is that the random seed means weak security, while the machine-id/ssh-key being duplicate mostly just means weird behaviour...

The random seed stuff is particularly relevant in virtualized environments, that have no proper entropy from anywhere else. Physical bare-metal systems do not need it at all, as they generally have hardware RNGs these days, and thus plenty entropy any time they want. However, virtualized environments in particular work with system images that are duplicated across nodes, and thus we need to be extra careful with that.

Also, I am note entirely sure that flushing out the random seed when duplicating the image is really the best idea, since it means that not just the many instances of one specific image share the same entropy pool at runtime, but in fact all instances of all images share it... I think there's some value in shipping a random seed in an image hence...

Merging this change as is at the minimum compat breakage: previously we didn't require people to flush out the random seed when duplicating images, with this change we suddenly would do. I am not sure I like such a compat breakage.

hence, I am still sure: this should be opt-in via a kernel cmdline option, and not the default. And the relationship to duplicated system images must be clearly documented.

**Comment by keszybz** [comment]
> Sure, but who actually does that? And if they do, who'd think about random-seed?

I think people generally know about this, for example the Fedora kickstarts remove machine-id:
https://pagure.io/fedora-kickstarts/blob/master/f/fedora-docker-base.ks#_91

But you're right, nobody seems to care about random-seed.

But maybe we could piggy-back on the removal of random-seed, and do the following: when storing the random-seed, also store the hostname and machine-id in that file. On bootup, if the hostname is localhost, or the hostname is different than what we stored, or if the machine-id is different than what we stored, do not credit the entropy (i.e. write to /dev/urandom as before). If those criteria are met however, credit the entropy (i.e. do what the patch does). ?

**Comment by poettering** [comment]
@keszybz hmm, not a fan of too many automatisms like that, but I figure that sounds acceptible. But I'd still would like to see a boot time option for this, too.

**Comment by bitstreamout** [comment]
Thanks for the feedback and your requested changes.  Just finished these changes

**Comment by poettering** [comment]
Still, this should not happen unconditionally, see discussion above. Crediting entropy should only happen when we know its safe, see the discussion above.

**Comment by poettering** [comment]
I will close this now. We can't really solve this properly I fear, we never know where the saved randomness comes from, and whether it is used on a single system only. Sorry.

**Comment by cgwalters** [comment]
> as it is systematically incompatible with booting systems from identical images. 

I disagree - people building such images should be cleaning the random seed.

> But you're right, nobody seems to care about random-seed.

Well, that's not true; from the same repo:
https://pagure.io/fedora-kickstarts/blob/master/f/fedora-atomic.ks#_125

And further in fact:
https://github.com/projectatomic/rpm-ostree/blob/e94f8c9b5f8e327c792156d6f154aedcba96ef16/src/libpriv/rpmostree-postprocess.c#L589

rpm-ostree gets this *correct* and has done for a long time now.
And that line links to the bug that added it to `%post` - and [see my complaint there](https://bugzilla.redhat.com/show_bug.cgi?id=789407#c2

So...this is basically "let's punish everyone since we can't know which image-building tools are good".

Anyways here's my strawman proposal: let's do `RNDADDENTROPY` by default for `ConditionVirtualization=!true`.  People doing *bare metal* imaging is a far, far smaller set than e.g. people who boot a VM in AWS, add some packages/containers, and then snapshot it and use it to run many VMs.

Concretely this issue affects my home (baremetal) server which is too old to have `RDRAND`.

**Comment by poettering** [comment]
Another PR, very similar to this one just got opened, in #10621. In contrast to this PR it makes the logic opt-in through a cmdline switch. I am much more positive about that (though I think it should be an env var, see my comments on that PR), simply because it means people have to explicitly opt-in into the entropy crediting logic by adding a drop-in file for systemd-random-seed.service. This way we don't just silently and carelessly credit entropy in all cases, in particular golden images where we really shouldn't.

**Comment by poettering** [comment]
> I disagree - people building such images should be cleaning the random seed.

Well, but i am very sure that is not what people do who just use `debootstrap` or a similar tool to quickly put together an image to deploy.

**Comment by poettering** [comment]
> Anyways here's my strawman proposal: let's do `RNDADDENTROPY` by default for `ConditionVirtualization=!true`. People doing _bare metal_ imaging is a far, far smaller set than e.g. people who boot a VM in AWS, add some packages/containers, and then snapshot it and use it to run many VMs.

Well, I am not convinced that such sloppy solutions are a good idea. People build embedded devices with systemd, and we shouldn't blindly make there systems less secure...

Anyway, I figure that the proposed PR #10621 which makes the crediting optional should make you happy. It would mean that if you know exactly your usecase and that people won't duplicate your images after booting them once, then you can just add the drop-in to your system and entropy will always be credited... But this is only safe if you know exactly how people use your stuff, and that noone copies images blindly after booting them at least once...

**Comment by cgwalters** [comment]
> Anyway, I figure that the proposed PR #10621 which makes the crediting optional should make you happy. It would mean that if you know exactly your usecase and that people won't duplicate your images after booting them once, then you can just add the drop-in to your system and entropy will always be credited... 

Yeah, that sounds fine to me.

**Pull request closed without merging**